torchx/schedulers/ray_scheduler.py

```diff
@@ -0,0 +1,299 @@
+            required=True,
+            help="Use CLUSTER_CONFIG_FILE to access or create the Ray cluster.",
+        )
+        opts.add(
+            "cluster_name",
+            type_=str,
+            help="Override the configured cluster name.",
+        )
+        opts.add(
+            "start_cluster",
+            type_=bool,
+            default=False,
+            help="Start the cluster if needed.",
+        )
+        opts.add(
+            "copy_scripts",
+            type_=bool,
+            default=False,
+            help="Copy the script(s) to the cluster. Supported only for PyTorch DDP jobs.",
+        )
+        opts.add(
+            "copy_script_dirs",
+            type_=bool,
+            default=False,
+            help="Copy the directories containing the scripts to the cluster. "
+            "Supported only for PyTorch DDP jobs.",
+        )
+        opts.add(
+            "verbose",
+            type_=bool,
+            default=False,
+            help="Enable verbose output.",
+        )
+        return opts
+
+    def schedule(self, dryrun_info: AppDryRunInfo[RayJob]) -> str:
+        cfg: RayJob = dryrun_info.request
+
+        # 1. Start the cluster using `ray up` if requested by `cfg.start_cluster`.
+        #
+        # 2. Copy `cfg.scripts` or their directories to `~/<cfg.app_id>` on the
+        #    cluster using `ray rsync-up` if requested by `cfg.copy_scripts` or
+        #    `cfg.copy_script_dirs`.
+        #
+        # 3. Dump `cfg.actors` to a JSON file (e.g. `/tmp/<cfg.app_id>-actors.json`)
+        #    and copy to `~/<cfg.app_id>` on the cluster using `ray rsync-up`.
+        #
+        # 4. Copy `ray_main.py` which handles the actual job execution to
+        #    `~/<cfg.app_id>` on the cluster using `ray rsync-up`.
+        #
+        # 5. Start `ray_main.py` as a background process on the head node using
+        #    `ray exec`.
+
+        return cfg.app_id
+
+    def _submit_dryrun(self, app: AppDef, cfg: RunConfig) -> AppDryRunInfo[RayJob]:
+        app_id = make_unique(app.name)
+
+        cluster_cfg = cfg.get("cluster_config_file")
+        if not isinstance(cluster_cfg, str) or not os.path.isfile(cluster_cfg):
+            raise ValueError("The cluster configuration file must be a YAML file.")
+
+        job = RayJob(app_id, cluster_cfg)
+
+        def set_job_attr(cfg_name: str, cfg_type: Type) -> None:
+            cfg_value = cfg.get(cfg_name)
+            if cfg_value is None:
+                return
+
+            if not isinstance(cfg_value, cfg_type):
+                raise TypeError(
+                    f"The configuration value '{cfg_name}' must be of type {cfg_type.__name__}."
+                )
+
+            setattr(job, cfg_name, cfg_value)
+
+        set_job_attr("cluster_name", str)
+        set_job_attr("start_cluster", bool)
+        set_job_attr("copy_scripts", bool)
+        set_job_attr("copy_script_dirs", bool)
+        set_job_attr("verbose", bool)
```

Comment:
generally good to just set the arguments since this is harder to handle with Pyre. Runopts I believe already enforces types (if it doesn't we should probably merge that checkin in there anyways)

Reply:
An alternative is setting up a CLI with https://typer.tiangolo.com/ - I'm using it for torchprep and it's an absolute pleasure to use